File: app/src/main/java/com/quran/labs/androidquran/ui/helpers/AyahHighlight.java
package com.quran.labs.androidquran.ui.helpers;

import androidx.annotation.NonNull;

import java.util.HashSet;
import java.util.Set;

public class AyahHighlight {
  private String key;
  private boolean transition;

  public AyahHighlight(String key) {
    this.key = key;
    this.transition = false;
  }

  public AyahHighlight(String key, boolean transition) {
    this.key = key;
    this.transition = transition;
  }

  public String getKey() {
    return key;
  }

  public boolean isTransition() {
    return transition;
  }

  @Override
  public boolean equals(Object obj) {
    if (obj == this) {
      return true;
    }

    if((obj == null) || (obj.getClass() != this.getClass())) {
      return false;
    }

    AyahHighlight ayahHighlight = (AyahHighlight)obj;
    return this.key.equals(ayahHighlight.key);
  }

  @Override
  public int hashCode() {
    return key.hashCode();
  }

  @NonNull
  @Override
  public String toString() {
    return key;
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/helpers/AyahHighlight.kt
package com.quran.labs.androidquran.ui.helpers

sealed class AyahHighlight {

  abstract val key: String

  fun isTransition() = this is TransitionAyahHighlight

  data class SingleAyahHighlight(override val key: String) : AyahHighlight() {
    constructor(surah: Int, ayah: Int) : this("$surah:$ayah")
    companion object {
      @JvmStatic
      fun createSet(ayahKeys: Set<String>): Set<AyahHighlight> {
        return ayahKeys.map { SingleAyahHighlight(it) }.toSet()
      }
    }
  }

  data class TransitionAyahHighlight(
    val source: AyahHighlight,
    val destination: AyahHighlight
  ) : AyahHighlight() {
    override val key = "${source.key}->${destination.key}"
  }

}
